core/src/core/transaction.rs

```diff
@@ -379,35 +394,108 @@ impl Transaction {
 	}
 }
 
-/// A transaction input, mostly a reference to an output being spent by the
-/// transaction.
-#[derive(Debug, Copy, Clone)]
-pub struct Input(pub Commitment);
+/// A transaction input.
+///
+/// Primarily a reference to an output being spent by the transaction.
+/// But also information required to verify coinbase maturity through
+/// the lock_height hashed in the switch_commit_hash.
+#[derive(Debug, Clone, Copy)]
+pub struct Input{
+	commit: Commitment,
+	// We need to provide the switch_commit (switch_commit_hash preimage)
+	// to verify coinbase maturity.
+	switch_commit: Commitment,
+	// We need to provide lock_height to verify coinbase maturity.
+	lock_height: u64,
+}
 
 hashable_ord!(Input);
 
 /// Implementation of Writeable for a transaction Input, defines how to write
 /// an Input as binary.
 impl Writeable for Input {
 	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ser::Error> {
-		writer.write_fixed_bytes(&self.0)
+		writer.write_fixed_bytes(&self.commit)?;
+		writer.write_fixed_bytes(&self.switch_commit)?;
```

Comment:
Thought the switch commitment would be required in inputs only for coinbase transactions (with a non-zero lock height)?

Reply:
Are you proposing serializing Inputs differently for spending coinbase outputs vs. non-coinbase outputs?
I think we'd need to include the `output_features` (or a `is_coinbase` flag or something) in the input to do this (so we know what to expect during de-serialization)?

Reply:
Yes, that's what I'm proposing